README.md
# bps-lib-pushover

This is fork of https://github.com/strongloop-forks/strong-fork-pushover

Primary repo: https://github.com/substack/pushover 

Serve up git repositories over http and accept git pushes.

This library makes it super easy to set up custom git push deploy logic.

# example

``` js
var pushover = require('marlic-fork-pushover');
var repos = pushover('/tmp/repos');

repos.on('push', function (push) {
    console.log('push ' + push.repo + '/' + push.commit
        + ' (' + push.branch + ')'
    );
    push.accept();
});

repos.on('fetch', function (fetch) {
    console.log('fetch ' + fetch.commit);
    fetch.accept();
});

var http = require('http');
var server = http.createServer(function (req, res) {
    repos.handle(req, res);
});
server.listen(7000);
```

then start up the pushover server...

```
$ node example/simple.js
```

meanwhile...

```
$ git push http://localhost:7000/beep master
Counting objects: 356, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (133/133), done.
Writing objects: 100% (356/356), 46.20 KiB, done.
Total 356 (delta 210), reused 355 (delta 210)
To http://localhost:7000/beep
 * [new branch]      master -> master

```

and then...

```
$ node example/simple.js
push beep.git/d5013a53a0e139804e729a12107fc212f11e64c3 (master)
```

or...

```
$ git clone http://localhost:7000/beep.git
```

and then...

```
fetch beep.git/d5013a53a0e139804e729a12107fc212f11e64c3
```

# methods

var pushover = require('pushover')

## var repos = pushover(repoDir, opts={autoCreate:true})

Create a new repository collection from the directory `repoDir`.
`repoDir` should be entirely empty except for git repo directories.

If `repoDir` is a function, `repoDir(repo)` will be used to dynamically resolve
project directories. The return value of `repoDir(repo)` should be a string path
specifying where to put the string `repo`. Make sure to return the same value
for `repo` every time since `repoDir(repo)` will be called multiple times.

The return value, `repos` is an EventEmitter that emits the events listed below
in the events section.

By default, repository targets will be created if they don't exist. You can
disable that behavior with `opts.autoCreate`.

If `opts.checkout` is true, create and expected checked-out repos instead of
bare repos.

## repos.handle(req, res)

Handle incoming HTTP requests with a connect-style middleware.

Everything is admin-party by default.
Check the credentials further up the stack using basic auth or whatevs.

## repos.create(repoName, cb)

Create a new bare repository `repoName` in the instance repository directory.

Optionally get a callback `cb(err)` to be notified when the repository was
created.

## repos.mkdir(dir, cb)

Create a subdirectory `dir` in the repo dir with an errback `cb(err)`.

## repos.list(cb)

Get a list of all the repositories in the callback `cb(err, repos)`.

## repos.exists(repoName, cb)

Find out whether `repoName` exists in the callback `cb(exists)`.

# events

## repos.on('push', function (push) { ... }

Emitted when somebody does a `git push` to the repo.

Exactly one listener must call `push.accept()` or `push.reject()`. If there are
no listeners, `push.accept()` is called automatically.

`push` is an http duplex object (see below) with these extra properties:

* push.repo
* push.commit
* push.branch

## repos.on('tag', function (tag) { ... }

Emitted when somebody does a `git push --tags` to the repo.

Exactly one listener must call `tag.accept()` or `tag.reject()`. If there are
no listeners, `tag.accept()` is called automatically.

`tag` is an http duplex object (see below) with these extra properties:

* tag.repo
* tag.commit
* tag.version

## repos.on('fetch', function (fetch) { ... }

Emitted when somebody does a `git fetch` to the repo (which happens whenever you
do a `git pull` or a `git clone`).

Exactly one listener must call `fetch.accept()` or `fetch.reject()`. If there are
no listeners, `fetch.accept()` is called automatically.

`fetch` is an http duplex objects (see below) with these extra properties:

* fetch.repo
* fetch.commit

## repos.on('info', function (info) { ... }

Emitted when the repo is queried for info before doing other commands.

Exactly one listener must call `info.accept()` or `info.reject()`. If there are
no listeners, `info.accept()` is called automatically.

`info` is an http duplex object (see below) with these extra properties:

* info.repo

## repos.on('head', function (head) { ... }

Emitted when the repo is queried for HEAD before doing other commands.

Exactly one listener must call `head.accept()` or `head.reject()`. If there are
no listeners, `head.accept()` is called automatically.

`head` is an http duplex object (see below) with these extra properties:

* head.repo

## push.on('response', function(response, done) { ... })

Emitted when pushover creates a resposne stream that will be sent to the git client on the other end.

This should really only be used if you want to send verbose or error messages to the remote git client.

`response` is a writable stream that can accept buffers containing git packfile sidechannel transfer protocol encoded strings. `done` is a callback that must be called when you want to end the response.

If you create a response listener then you must either call the `done` function or execute the following end sequence when you want to end the response:

```js
response.queue(new Buffer('0000'))
response.queue(null)
```

If you never use the response event then the above data will be sent by default. Binding a listener to the response event will prevent the end sequence those from being sent, so you must send them yourself after sending any other messages.

# http duplex objects

The arguments to each of the events `'push'`, `'fetch'`, `'info'`, and `'head'`
are [http duplex](http://github.com/substack/http-duplex) that act as both http
server request and http server response objects so you can pipe to and from them.

For every event if there are no listeners `dup.accept()` will be called
automatically.

## dup.accept()

Accept the pending request.

## dup.reject()

Reject the pending request.

# install

With [npm](https://npmjs.org) do:

```
npm install pushover
```

# license

MIT

# kudos

Reading through
[grack](https://github.com/schacon/grack/blob/master/lib/git_http.rb)
was super handy.


CHANGES.md

n.n.n / 2019-10-27
==================



1.3.10 / 2019-10-27
===================

  * send reject message to git client in understandable for client format
  * add reference to req and res object for all event context object (Service)
  * test: remove ".git" from repo url
  * fixes from martindale/pushover: Eliminate .git in local filesystems
  * fixes from vdemedes/pushover: fix push handling; fix large data transfers; add missing require() for fs
  * Fork as marlic-fork-pushover; Revert "Describe fork in README"; add .gitignore
  * Merge tag 'v1.3.9'

v1.3.9 / 2015-01-21
===================

  * Merge branch 'release/1.3.9' into production
  * v1.3.9
  * Remove HttpDuplex from Service prototype
  * Merge tag 'v1.3.8'

v1.3.8 / 2015-01-13
===================

  * Merge branch 'release/1.3.8' into production
  * v1.3.8
  * Merge tag 'v1.3.7'

v1.3.7 / 2015-01-13
===================

  * Merge branch 'release/1.3.7' into production
  * v1.3.7
  * Describe fork in README
  * Use 'git service' instead of 'git-service'
  * tests: use cross-platform tmpdir
  * tests: use 127.0.0.1 instead of localhost
  * test: improve reject test wording
  * fix repo name in tests
  * test: fix path/fs.exists reference
  * travis: update node versions
  * Use 'git command' instead of 'git-command'
  * Fork as strong-fork-pushover

1.3.6 / 2014-05-17
==================

  * 1.3.6
  * self

1.3.5 / 2014-04-22
==================

  * 1.3.5
  * print errors in the reject test
  * wrap spawn calls in better error reporting
  * inherits package instead
  * indent
  * s/\t/\s/g
  * s/\t/\s/g
  * Added gzip decoding when git fetches pack using gzip compression
  * add done callback and document the response event
  * add response stream for controlling what gets written to the git client response

1.3.4 / 2013-12-11
==================

  * 1.3.4
  * Using util.inherits instead of new EventEmitter and calling contstructor. Otherwise multiple pushover instances uses one EventEmitter instance

1.3.3 / 2013-04-07
==================

  * Merge branch 'master' of github.com:substack/pushover
  * add back http-duplex into service responses
  * Merge pull request #20 from mantoni/patch-1
  * update the dirmap test to work with auto .git inclusion
  * Corrected port number in example

1.3.2 / 2013-04-06
==================

  * 1.3.2
  * using a regex instead of substr
  * Merge branch 'master' of git://github.com/MRdNk/pushover into MRdNk-master
  * Repo name: If repo does not end with .git - then add .git to the repo name

1.3.1 / 2013-04-06
==================

  * `connection: close` fixes everything; 1.3.1

1.3.0 / 2013-04-05
==================

  * 1.3.0
  * use .once()
  * using through in lib/service to get proper backpressure
  * pipe stdout directly through instead of using a custom onexit handler
  * readme art

1.2.1 / 2013-02-04
==================

  * fix for some repos that just send 0000 on push-receive for some reason

1.2.0 / 2012-12-21
==================

  * 1.2.0
  * document dirmap
  * dirmap test finally passes; thread the functionality through repos(dirmap) instead
  * failing dirmap test

1.1.1 / 2012-12-12
==================

  * fix race condition resulting in lost chunks and hanging git pushes
  * Merge branch 'master' of git://github.com/distracteddev/pushover
  * Fix upload-pack regex
  * Fix upload-pack regex

1.1.0 / 2012-10-14
==================

  * 1.1.0
  * node_modules should never be in .gitignore
  * emit multiple events if commits and tags are pushed
  * handle tags

1.0.3 / 2012-09-26
==================

  * 1.0.3
  * Fix crash on some requests.

1.0.2 / 2012-09-23
==================

  * make missing directories in create()

1.0.1 / 2012-09-22
==================

  * subdirs work
  * failing test for subdirs

1.0.0 / 2012-09-22
==================

  * s/qs/querystring/
  * update readme, example with all the new events
  * add cwd and repo to info and head duplex event objects
  * head events
  * info requests split out into a separate lib file
  * info events, first half
  * all tests pass again, "fetch" event
  * reject test passes with the new accept/reject event api
  * fix test to compensate for repos.create()
  * refactor service handling logic into lib/service.js, launch the command after emitting and parsing the values
  * make sure the reject test directory never gets created, currently fails
  * remove repos.listen()
  * split up big file into some littler files
  * fix so all tests pass again
  * passing reject test
  * switched over to http-duplex api
  * update tests for http-duplex usage
  * passing the new tests
  * updated tests for new push signature

0.1.6 / 2012-08-14
==================

  * 0.1.6, document branch argument to push event
  * push emits branch, passes branch test
  * failing test to get branch data

0.1.5 / 2012-08-13
==================

  * fs.exists
  * documented "push", repo, commit
  * emit the commit hash as the second arg, passes its test
  * failing test to check the commit hash as the second arg from "push" events

0.1.4 / 2012-07-22
==================

  * fix all the race conditions
  * fix race condition in node 0.8. due to "exit"/end event ordering

0.1.3 / 2012-07-07
==================

  * seq is still a dev dep
  * fs.exists
  * remove unnecessary dep, bump travis to 0.8
  * travis badge

0.1.2 / 2012-02-28
==================

  * bump and using travis
  * Bubble up listening event to pushover.

0.1.1 / 2012-01-14
==================

  * bump for opts.checkout
  * minor style changes
  * self not defined
  * Add option to work with checkouts rather than bare repos.

0.1.0 / 2011-11-23
==================

  * readme notes and a bump for autocreate
  * autoCreate test passes
  * create() is async, check for autoCreate option
  * handle now ensures the repository exists and creates a bare one if required. Effectively removes the need for explicit create call

0.0.0 / 2011-11-17
==================

  * note about repos as an event emitter

test/repos.js
var test = require('tap').test;
var pushover = require('../');

var fs = require('fs');
var path = require('path');
var tmpdir = require('os').tmpdir;
var exists = fs.exists || path.exists;

var spawn = require('child_process').spawn;
var exec = require('child_process').exec;
var http = require('http');

var seq = require('seq');

test('create, push to, and clone a repo', function (t) {
    t.plan(17);
    
    var repoDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var srcDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var dstDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var lastCommit;
    
    fs.mkdirSync(repoDir, 0700);
    fs.mkdirSync(srcDir, 0700);
    fs.mkdirSync(dstDir, 0700);
    
    var repos = pushover(repoDir, { autoCreate : false });
    var port = Math.floor(Math.random() * ((1<<16) - 1e4)) + 1e4;
    var server = http.createServer(function (req, res) {
        repos.handle(req, res);
    });
    server.listen(port, '127.0.0.1');
    
    process.chdir(srcDir);
    seq()
        .seq(function () { repos.create('doom', this) })
        .seq(function () {
            var ps = spawn('git', [ 'init' ]);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', this.ok);
        })
        .seq(function () {
            fs.writeFile(srcDir + '/a.txt', 'abcd', this);
        })
        .seq(function () {
            spawn('git', [ 'add', 'a.txt' ]).on('exit', this.ok)
        })
        .seq_(function (next) {
            var ps = spawn('git', [ 'commit', '-am', 'a!!' ]);
            ps.on('exit', this.ok);
            ps.stdout.pipe(process.stdout, { end : false });
        })
        .seq(function () {
            var ps = spawn('git', ['tag', '0.0.1']);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', this.ok);
        })
        .seq(function () {
            fs.writeFile(srcDir + '/a.txt', 'efgh', this);
        })
        .seq(function () {
            spawn('git', [ 'add', 'a.txt' ]).on('exit', this.ok)
        })
        .seq_(function (next) {
            var ps = spawn('git', [ 'commit', '-am', 'a!!' ]);
            ps.on('exit', function () {
                exec('git log | head -n1', function (err, stdout) {
                    lastCommit = stdout.split(/\s+/)[1];
                    next();
                });
            });
            ps.stdout.pipe(process.stdout, { end : false });
        })
        .seq(function () {
            var ps = spawn('git', ['tag', '0.0.2']);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', this.ok);
        })
        .seq(function () {
            var ps = spawn('git', [
                'push', '--tags', 'http://127.0.0.1:' + port + '/doom', 'master'
            ]);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', this.ok);
        })
        .seq(function () {
            process.chdir(dstDir);
            spawn('git', [ 'clone', 'http://127.0.0.1:' + port + '/doom' ])
                .on('exit', this.ok)
        })
        .seq_(function (next) {
            exists(dstDir + '/doom/a.txt', function (ex) {
                t.ok(ex, 'a.txt exists');
                next();
            })
        })
        .seq(function () {
            server.close();
            t.end();
        })
        .catch(t.fail)
    ;
    
    repos.on('push', function (push) {
        t.equal(push.repo, 'doom', 'repo name');
        t.equal(push.commit, lastCommit, 'commit ok');
        t.equal(push.branch, 'master', 'master branch');
        
        t.equal(push.headers.host, '127.0.0.1:' + port, 'http host');
        t.equal(push.method, 'POST', 'is a post');
        t.equal(push.url, '/doom/git-receive-pack', 'receive pack');
        
        push.accept();
    });

    var firstTag = true;
    repos.on('tag', function (tag) {
        t.equal(tag.repo, 'doom', 'repo name');
        t.equal(tag.version, '0.0.' + (firstTag? 1 : 2), 'tag received');

        t.equal(tag.headers.host, '127.0.0.1:' + port, 'http host');
        t.equal(tag.method, 'POST', 'is a post');
        t.equal(tag.url, '/doom/git-receive-pack', 'receive pack');

        tag.accept();
        firstTag = false;
    });
});


test/dirmap.js
var test = require('tap').test;
var pushover = require('../');

var fs = require('fs');
var path = require('path');
var tmpdir = require('os').tmpdir;
var exists = fs.exists || path.exists;

var spawn = require('child_process').spawn;
var http = require('http');

var seq = require('seq');

var repoDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
var srcDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
var dstDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
var targetDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));

fs.mkdirSync(repoDir, 0700);
fs.mkdirSync(srcDir, 0700);
fs.mkdirSync(dstDir, 0700);
fs.mkdirSync(targetDir, 0700);
    
var repos;
var server = http.createServer(function (req, res) {
    repos.handle(req, res);
});

test(function (t) {
    server.listen(0, '127.0.0.1', function () {
        setTimeout(t.end.bind(t), 1000);
    });
});

test('clone into programatic directories', function (t) {
    t.plan(9);
    
    repos = pushover(function (dir) {
        t.equal(dir, 'doom');
        return path.join(targetDir, dir);
    });
    var port = server.address().port;
    
    process.chdir(srcDir);
    seq()
        .seq(function () {
            var ps = spawn('git', [ 'init' ]);
            ps.on('exit', this.ok);
        })
        .seq(function () {
            fs.writeFile(srcDir + '/a.txt', 'abcd', this);
        })
        .seq(function () {
            spawn('git', [ 'add', 'a.txt' ]).on('exit', this.ok)
        })
        .seq(function () {
            spawn('git', [ 'commit', '-am', 'a!!' ]).on('exit', this.ok)
        })
        .seq(function () {
            var ps = spawn('git', [
                'push', 'http://127.0.0.1:' + port + '/doom', 'master'
            ]);
            ps.on('exit', this.ok);
            ps.on('exit', function (code) {
                t.equal(code, 0);
            });
        })
        .seq(function () {
            process.chdir(dstDir);
            spawn('git', [ 'clone', 'http://127.0.0.1:' + port + '/doom' ])
                .on('exit', this.ok)
        })
        .seq_(function (next) {
            exists(dstDir + '/doom/a.txt', function (ex) {
                t.ok(ex, 'a.txt exists');
                next();
            })
        })
        .seq_(function (next) {
            exists(targetDir + '/HEAD', function (ex) {
                t.ok(ex, 'INFO exists');
                next();
            })
        })
        .catch(t.fail)
    ;
    
    repos.on('push', function (push) {
        t.equal(push.repo, 'doom');
        push.accept();
    });
    
    t.on('end', function () {
        server.close();
    });
});


test/reject.js
var test = require('tap').test;
var pushover = require('../');

var fs = require('fs');
var path = require('path');
var tmpdir = require('os').tmpdir;
var exists = fs.exists || path.exists;

var spawn_ = require('child_process').spawn;
function spawn (cmd, args, opts) {
    var ps = spawn_(cmd, args, opts);
    ps.on('error', function (err) {
        console.error(
            err.message + ' while executing: '
            + cmd + ' ' + args.join(' ')
        );
    });
    return ps;
}
var exec = require('child_process').exec;
var http = require('http');

var seq = require('seq');

test('create, push to, and clone a repo', function (t) {
    t.plan(9);
    
    var repoDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var srcDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var dstDir = path.join(tmpdir(), Math.floor(Math.random() * (1<<30)).toString(16));
    var lastCommit;
    
    fs.mkdirSync(repoDir, 0700);
    fs.mkdirSync(srcDir, 0700);
    fs.mkdirSync(dstDir, 0700);
    
    var repos = pushover(repoDir, { autoCreate : false });
    var port = Math.floor(Math.random() * ((1<<16) - 1e4)) + 1e4;
    var server = http.createServer(function (req, res) {
        repos.handle(req, res);
    });
    server.listen(port, '127.0.0.1');
    
    t.on('end', function () {
        server.close();
    });
    
    process.chdir(srcDir);
    seq()
        .seq(function () { repos.create('doom', this) })
        .seq(function () {
            var ps = spawn('git', [ 'init' ]);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', this.ok);
        })
        .seq(function () {
            fs.writeFile(srcDir + '/a.txt', 'abcd', this);
        })
        .seq(function () {
            spawn('git', [ 'add', 'a.txt' ]).on('exit', this.ok)
        })
        .seq_(function (next) {
            var ps = spawn('git', [ 'commit', '-am', 'a!!' ]);
            ps.on('exit', function () {
                exec('git log | head -n1', function (err, stdout) {
                    lastCommit = stdout.split(/\s+/)[1];
                    next();
                });
            });
            ps.stdout.pipe(process.stdout, { end : false });
        })
        .seq_(function (next) {
            var ps = spawn('git', [
                'push', 'http://127.0.0.1:' + port + '/doom', 'master'
            ]);
            ps.stderr.pipe(process.stderr, { end : false });
            ps.on('exit', function (code) {
                t.notEqual(code, 0);
                next();
            });
        })
        .seq(setTimeout, seq, 1000)
        .seq(function () {
            var glog = spawn('git', [ 'log' ], { cwd : repoDir + '/doom' });
            glog.on('exit', function (code) {
                t.notEqual(code, 0);
            });
            var data = '';
            glog.stderr.on('data', function (buf) { data += buf });
            glog.stderr.on('end', function (buf) {
                t.ok(/bad default revision 'HEAD'/.test(data), 'bad default revision');
            });
        })
        .catch(t.fail)
    ;
    
    repos.on('push', function (push) {
        t.equal(push.repo, 'doom', 'repo name');
        t.equal(push.commit, lastCommit, 'commit ok');
        t.equal(push.branch, 'master', 'master branch');
        
        t.equal(push.headers.host, '127.0.0.1:' + port, 'http host');
        t.equal(push.method, 'POST', 'is a post');
        t.equal(push.url, '/doom/git-receive-pack', 'receive pack');
        
        push.reject(500, 'ACCESS DENIED');
    });
});


index.js
var fs = require('fs');
var path = require('path');
var http = require('http');
var mkdirp = require('mkdirp');
var inherits = require('inherits');

var spawn = require('child_process').spawn;
var EventEmitter = require('events').EventEmitter;

var onexit = require('./lib/onexit');

module.exports = function (repoDir, opts) {
    if (!opts) opts = {};
    var dirMap = typeof repoDir === 'function'
        ? repoDir
        : function (dir) { return path.join(repoDir, dir) }
    ;
    return new Git(dirMap, opts);
};

function Git (dirMap, opts) {
    EventEmitter.call(this);

    this.dirMap = dirMap;
    this.autoCreate = opts.autoCreate === false ? false : true;
    this.checkout = opts.checkout;
}

inherits(Git, EventEmitter);

Git.prototype.list = function (cb) {
    fs.readdir(this.dirMap(), cb);
};

Git.prototype.exists = function (repo, cb) {
    (fs.exists || path.exists)(this.dirMap(repo), cb);
};

Git.prototype.mkdir = function (dir, cb) {
    mkdirp(this.dirMap(dir), cb);
};

Git.prototype.create = function (repo, cb) {
    var self = this;
    if (typeof cb !== 'function') cb = function () {};

    self.exists(repo, function (ex) {
        if (!ex) self.mkdir(repo, next)
        else next()
    });
    
    function next (err) {
        if (err) return cb(err);
        
        var dir = self.dirMap(repo);
        if (self.checkout) {
            var ps = spawn('git', [ 'init', dir ]);
        }
        else {
            var ps = spawn('git', [ 'init', '--bare', dir ]);
        }
        
        var err = '';
        ps.stderr.on('data', function (buf) { err += buf });
        
        onexit(ps, function (code) {
            if (!cb) {}
            else if (code) cb(err || true)
            else cb(null)
        });
    }
};

Git.prototype.handle = require('./lib/handle');


lib/service.js
var spawn = require('child_process').spawn;
var through = require('through');
var inherits = require('inherits');
var EventEmitter = require('events').EventEmitter;
var zlib = require('zlib');
var noCache = require('./no_cache');

var encodings = {
    'gzip': function() {return zlib.createGunzip();},
    'deflate': function() {return zlib.createDeflate();}
};

module.exports = function (opts, req, res) {
    var service = new Service(opts, req, res);
    
    Object.keys(opts).forEach(function (key) {
        service[key] = opts[key];
    });
    return service;
};

var headerRE = {
    'receive-pack' : '([0-9a-fA-F]+) ([0-9a-fA-F]+)'
        + ' refs\/(heads|tags)\/(.*?)( |00|\u0000)'
        + '|^(0000)$'
    ,
    'upload-pack' : '^\\S+ ([0-9a-fA-F]+)'
};

inherits(Service, EventEmitter);

function Service (opts, req, res) {
    EventEmitter.call(this);
    var self = this;
    
    self.headers = req.headers;
    self.method = req.method;
    self.url = req.url;
    self.req = req;
    self.res = res;

    self.status = 'pending';
    self.repo = opts.repo;
    self.service = opts.service;
    self.cwd = opts.cwd;
    
    var buffered = through().pause();

    // stream needed to receive data after decoding, but before accepting
    var ts = through();

    var decoder = encodings[req.headers['content-encoding']];
    if (decoder) {
        // data is compressed with gzip or deflate
        req.pipe(decoder()).pipe(ts).pipe(buffered);
    } else {
        // data is not compressed
        req.pipe(ts).pipe(buffered);
    }
    
    var data = '';
    ts.once('data', function (buf) {
        data += buf;
        
        var ops = data.match(new RegExp(headerRE[self.service], 'gi'));
        if (!ops) return;
        data = undefined;
       
        ops.forEach(function(op) {
            var m = op.match(new RegExp(headerRE[self.service]));
            if (self.service === 'receive-pack') {
                self.last = m[1];
                self.commit = m[2];

                if (m[3] == 'heads') {
                    var type = 'branch';
                    self.evName = 'push';
                } else {
                    var type = 'version';
                    self.evName = 'tag';
                }

                var headers = {
                    last: self.last,
                    commit : self.commit
                };
                headers[type] = self[type] = m[4];
                self.emit('header', headers);
            }
            else if (self.service === 'upload-pack') {
                self.commit = m[1];
                self.evName = 'fetch';
                self.emit('header', { commit : self.commit });
            }
        });
    });
    
    self.once('accept', function () {
        process.nextTick(function () {
            var cmd = [ 'git', opts.service, '--stateless-rpc', opts.cwd ];
            var ps = spawn(cmd[0], cmd.slice(1));
            ps.on('error', function (err) {
                self.emit('error', new Error(
                    err.message + ' running command ' + cmd.join(' ')
                ));
            });
            
            self.emit('service', ps);
            
            var respStream = through(function(c) {
                if (self.listeners('response').length === 0)
                    return this.queue(c);
                // prevent git from sending the close signal
                if (c.length === 4 && c.toString() === '0000')
                    return;

                this.queue(c);
            }, function() {
                if (self.listeners('response').length > 0)
                    return;
                this.queue(null);
            }, {
                autoDestroy: false
            });
            
            function endResponse() {
                respStream.queue(new Buffer('0000'));
                respStream.queue(null);

                self.emit('exit');
            }

            ps.stdout.pipe(respStream).pipe(res);
            
            buffered.pipe(ps.stdin);
            buffered.resume();

            ps.on('exit', function () {
                if (self.listeners('response').length === 0) {
                    self.emit('exit');

                    return;
                }

                self.emit('response', respStream, endResponse);
            });
        });
    });
    
    self.once('reject', function (code, msg) {

        res.setHeader('content-type', 'application/x-git-' + self.service + '-result');
        noCache(res);
        res.end(gitError(`error: ${msg}`));
    });
}

Service.prototype.accept = function (dir) {
    if (this.status !== 'pending') return;
    
    this.status = 'accepted';
    this.emit('accept', dir);
};

Service.prototype.reject = function (code, msg) {
    if (this.status !== 'pending') return;
    
    if (msg === undefined && typeof code === 'string') {
        msg = code;
        code = 500;
    }
    this.status = 'rejected';
    this.emit('reject', code || 500, msg);
};

function gitError(msg) {
    const lengthDec = Buffer.from(msg).length + 6;
    const lengthHex = lengthDec.toString(16).padStart(4, '0');
    const stx = String.fromCharCode(2);
    const etx = String.fromCharCode(3);
    const nul = String.fromCharCode(0);
    const cr = String.fromCharCode(13);
    const nl = String.fromCharCode(10);
    const end = `000e${etx}aborting${nul}${cr}${nl}`;
    return Buffer.from(`${lengthHex}${stx}${msg}${nl}${end}`);
}
